File: head_first_design_pattern/C++/strategy.cpp
//Strategy Pattern

#include<iostream>

using namespace::std;

// interface
class FlyBehavior
{
public:
	virtual void fly() = 0;
};

class QuackBehavior
{
public:
	virtual void quack() = 0;
};

//behavior class
class Fly:public FlyBehavior
{
	void fly()
	{
		cout << "Fly" << endl;
	}
};

class FlyNoWay :public FlyBehavior
{
	void fly()
	{
		cout << "I can't fly" << endl;
	}
};

class FlyRocketPowered :public FlyBehavior
{
	void fly()
	{
		cout << "I'm flying with a rocket!" << endl;
	}
};

class FlyWithWings :public FlyBehavior
{
	void fly()
	{
		cout << "I'm flying!!" << endl;
	}
};

class Quack :public QuackBehavior
{
	void quack() 
	{
		cout << "Quack" << endl;
	}
};

class MuteQuack : public QuackBehavior
{
	void quack()
	{
		cout << "<< Silence >>" << endl;
	}
};

class Squeak :public QuackBehavior
{
	void quack()
	{
		cout << "Squeak" << endl;
	}
};

class Duck
{
public:
	FlyBehavior *flyBehavior;
	QuackBehavior *quackBehavior;
	virtual void display() = 0;
	void swim()
	{
		cout << "All ducks float, even decoys!" << endl;
	}

	void perform_fly()
	{
		flyBehavior->fly();
	}
	
	void perform_quack()
	{
		quackBehavior->quack();
	}

	void set_fly_behavior(FlyBehavior *fb)
	{
		flyBehavior = fb;
	}

	void set_quack_behavior(QuackBehavior *qb)
	{
		quackBehavior = qb;
	}
};


class ModelDuck :public Duck
{
public:
	ModelDuck()
	{
		flyBehavior = new FlyNoWay();
		quackBehavior = new Quack();
	};

	void display()
	{
		cout << "I'm a model duck" << endl;
	}
};

int main(void)
{
	Duck *model = new ModelDuck();
	model->perform_fly();
	model->set_fly_behavior(new FlyRocketPowered());
	model->perform_fly();
	return 0;
}

File: head_first_design_pattern/Python/strategy.py
from abc import ABC, abstractmethod


# interface
class FlyBehavior(ABC):
    @abstractmethod
    def fly(self):
        pass


class QuackBehavior(ABC):
    @abstractmethod
    def quack(self):
        pass


# behavior
class Fly(FlyBehavior):
    def fly(self):
        print("Fly")


class FlyNoWay(FlyBehavior):
    def fly(self):
        print("I can't fly")


class FlyRocketPowered(FlyBehavior):
    def fly(self):
        print("I'm flying with a rocket!")


class Quack(QuackBehavior):
    def quack(self):
        print("Quack")


class MuteQuack(QuackBehavior):
    def quack(self):
        print("<< Silence >>")


class Squeak(QuackBehavior):
    def quack(self):
        print("Squeak")


class Duck(ABC):
    def __init__(self):
        pass

    @abstractmethod
    def display(self):
        pass

    def swim(self):
        print("All ducks float, even decoys!")

    def perform_fly(self):
        self.fly_behavior.fly()

    def perfrom_quack(self):
        self.quack_behavior.quack()

    def set_fly_behavior(self, fb):
        self.fly_behavior= fb

    def set_quack_behavior(self, qb):
        self.quack_behavior = qb


class ModelDuck(Duck):
    def __init__(self):
        self.fly_behavior = FlyNoWay()
        self.quack_behavior = Quack()
        pass

    def display(self):
        print("I'm a model duck")

model = ModelDuck()
model.perform_fly()
model.set_fly_behavior(FlyRocketPowered())
model.perform_fly()